### PR TITLE
Added Tiles view

### PR DESCRIPTION
**Pull request by konadrian**
Added Tiles view
After aprroved this PR required create another with sql changes
TilesView

### CODE REVIEW

File: languages/en-US/_Base.json [hunk 1]
```diff
@@ -1480,7 +1480,13 @@
...
     "LBL_GREATER_OR_EQUAL_FIELD": "greater than or equal to field (>=)",
     "LBL_DRAG_AND_DROP_FILES": "Drag & Drop your files or Browse",
-    "LBL_SELECT_CALENDAR": "Select calendar"
+    "LBL_SELECT_CALENDAR": "Select calendar",
+    "LBL_VIEW_TILES": "Tiles view",
+    "LBL_CHOOSE_TILE_SIZE": "Tile size",
+    "LBL_VERY_SMALL_TILE": "Very small",
+    "LBL_SMALL_TILE": "Small",
+    "LBL_MEDIUM_TILE": "Medium",
+    "LBL_BIG_TILE": "Big"
```
Comment (rskrzypczak) [inline languages/en-US/_Base.json]: LBL_TILES_...

File: layouts/basic/modules/Vtiger/TilesHeader.tpl [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+					</div>
+					<div class=" d-flex flex-nowrap u-overflow-scroll-non-desktop">
+						<label for="tiles-size" class="mr-1 col-form-label w-100">{\App\Language::translate('LBL_CHOOSE_TILE_SIZE', $MODULE_NAME)}</label>
+						<select name="tile-size" class="form-control js-tiles-size select2" id="tiles-size">
+							<option value="very-small" {if $TILE_SIZE eq 'very-small'} selected {/if}> {\App\Language::translate('LBL_VERY_SMALL_TILE', $MODULE_NAME)}</option>
+							<option value="small" {if $TILE_SIZE eq 'small'} selected {/if}> {\App\Language::translate('LBL_SMALL_TILE', $MODULE_NAME)}</option>
+							<option value="medium" {if $TILE_SIZE eq 'medium'} selected {/if}> {\App\Language::translate('LBL_MEDIUM_TILE', $MODULE_NAME)}</option>
+							<option value="big" {if $TILE_SIZE eq 'big'} selected {/if}> {\App\Language::translate('LBL_BIG_TILE', $MODULE_NAME)}</option>
+						</select>
+					</div>
```
Comment (rskrzypczak) [inline layouts/basic/modules/Vtiger/TilesHeader.tpl]: new template

File: layouts/basic/modules/Vtiger/TilesLeftSide.tpl [hunk 3]
```diff
@@ -0,0 +1,33 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 5.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
+	<!-- tpl-Base-TilesLeftSide -->
```
Comment (rskrzypczak) [inline layouts/basic/modules/Vtiger/TilesLeftSide.tpl]: file name

File: layouts/basic/modules/Vtiger/TilesPreProcess.tpl [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+						</div>
+					</div>
+				</div>
+			{/if}
+			<div class="contentsDiv{if !$BREADCRUMBS_ACTIVE || !$HEADER_LINKS['LIST_VIEW_HEADER']} pt-2{/if}">
+				<a class="btn btn-outline-dark d-md-none o-header-toggle__actions-btn js-header-toggle__actions-btn mb-1" href="#" data-js="click" role="button"
+					aria-expanded="false" aria-controls="o-view-actions__container">
+					<span class="fas fa-ellipsis-h fa-fw" title="{\App\Language::translate('LBL_ACTION_MENU')}"></span>
+				</a>
+				{include file=\App\Layout::getTemplatePath('TilesHeader.tpl', $MODULE)}
```
Comment (rskrzypczak) [inline layouts/basic/modules/Vtiger/TilesPreProcess.tpl]: up

File: modules/Vtiger/models/Module.php [hunk 5]
```diff
@@ -984,6 +984,14 @@ public function getSideBarLinks($linkParams)
 				'linkicon' => 'yfi yfi-kanban',
 			]);
 		}
+		if ($this->isPermitted('Tiles')) {
```
Comment (rskrzypczak) [inline modules/Vtiger/models/Module.php]:
```suggestion
		if ($this->isPermitted('TilesView')) {
```

File: modules/Vtiger/models/Record.php [hunk 6]
```diff
@@ -480,6 +480,30 @@ public function getListViewDisplayValue($field, $rawText = false)
...
+	{
+		if ($field instanceof Vtiger_Field_Model) {
+			if (!empty($field->get('source_field_name')) && isset($this->ext[$field->get('source_field_name')][$field->getModuleName()])) {
+				return $this->ext[$field->get('source_field_name')][$field->getModuleName()]->getListViewDisplayValue($field, $rawText);
+			}
+		} else {
+			$field = $this->getModule()->getFieldByName($field);
+		}
+		$uitypeModel = $field->getUITypeModel();
+		$uitypeModelMethod = method_exists($uitypeModel, 'getTilesDisplayValue') ? 'getTilesDisplayValue' : 'getListViewDisplayValue';
```
Comment (rskrzypczak) [inline modules/Vtiger/models/Record.php]: getTilesDisplayValue add to global base uitype model

File: layouts/basic/modules/Vtiger/TilesContents.tpl [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+														<small> {$LISTVIEW_ENTRY->getListViewDisplayValue($LISTVIEW_HEADER)} </small>
+													</a>
+												{else}
+													<small>
+														{if $FIELD_DATA_TYPE eq 'multiImage'}
+															{$LISTVIEW_ENTRY->getTilesDisplayValue($LISTVIEW_HEADER)}
+														{else}
+															{$LISTVIEW_ENTRY->getListViewDisplayValue($LISTVIEW_HEADER)}
+
+														{/if}
```
Comment (rskrzypczak) [inline layouts/basic/modules/Vtiger/TilesContents.tpl]:
```suggestion
           {$LISTVIEW_ENTRY->getTilesDisplayValue($LISTVIEW_HEADER)}
```

File: modules/Vtiger/uitypes/MultiImage.php [hunk 8]
```diff
@@ -440,4 +440,66 @@ public function uploadTempFile(array $files, int $recordId, ?string $hash = null
...
+	 * @param bool|int                 $record
+	 * @param bool|Vtiger_Record_Model $recordModel
+	 * @param bool                     $rawText
+	 *
+	 * @return void
+	 */
+	public function getTilesDisplayValue($value, $record = false, $recordModel = false, $rawText = false)
+	{
+		$value = \App\Json::decode($value);
+		if (!$value) {
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/MultiImage.php]:
```suggestion
		if (!$value || !($value = \App\Json::decode($value)) {
```

File: modules/Vtiger/uitypes/MultiImage.php [hunk 9]
```diff
@@ -440,4 +440,66 @@ public function uploadTempFile(array $files, int $recordId, ?string $hash = null
...
+			return '';
+		}
+		if (!$record && $recordModel) {
+			$record = $recordModel->getId();
+		}
+		$carouselId = "tileCarousel{$record}";
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/MultiImage.php]: \App\Layout::getUniqueId(.. ,$record.'IM-');

File: modules/Vtiger/views/Tiles.php [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+	protected $tilesViewHeaders = false;
+	/** @var Vtiger_ListView_Model List view model instance */
+	protected $listViewModel;
+	/** @var int|string List view name or id. */
+	protected $viewName;
+
+	/** {@inheritdoc} */
+	public function checkPermission(App\Request $request)
+	{
+		if (!Users_Privileges_Model::getCurrentUserPrivilegesModel()->hasModuleActionPermission($request->getModule(), 'Tiles')) {
```
Comment (rskrzypczak) [inline modules/Vtiger/views/Tiles.php]: check

File: modules/Vtiger/views/Tiles.php [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+ * @package View
+ *
+ * @copyright YetiForce S.A.
+ * @license   YetiForce Public License 5.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author    Adrian Kon <a.kon@yetiforce.com>
+ */
+/**
+ * Tiles view class.
+ */
+class Vtiger_Tiles_View extends Vtiger_Index_View
```
Comment (rskrzypczak) [inline modules/Vtiger/views/Tiles.php]: check extends Vtiger_List_view

File: modules/Vtiger/views/Tiles.php [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+	 * Function to initialize the required data in smarty to display the Tiles view contents.
+	 *
+	 * @param App\Request   $request
+	 * @param Vtiger_Viewer $viewer
+	 *
+	 * @return void
+	 */
+	public function initializeTilesViewContents(App\Request $request, Vtiger_Viewer $viewer): void
+	{
+		$moduleName = $request->getModule();
```
Comment (rskrzypczak) [inline modules/Vtiger/views/Tiles.php]: check parent::initialize..

File: modules/Vtiger/views/Tiles.php [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+				$_SESSION['lvs'][$moduleName]['entityState'] = $request->getByType('entityState');
+			}
+			$this->initializeTilesViewContents($request, $viewer);
+			$viewer->assign('USER_MODEL', Users_Record_Model::getCurrentUserModel());
+			$viewer->assign('MODULE_NAME', $moduleName);
+			$viewer->assign('MODULE_MODEL', Vtiger_Module_Model::getInstance($moduleName));
+			$viewer->assign('VIEWID', $this->viewName);
+		}
+		$viewer->assign('VIEW', $request->getByType('view', 1));
+		$viewer->view('TilesContents.tpl', $moduleName);
```
Comment (rskrzypczak) [inline modules/Vtiger/views/Tiles.php]: check

public function ..TplName(App\Request $request)
	{
		return 'TilesContents.tpl';
	}

File: modules/Vtiger/views/Tiles.php [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+	}
+
+	/** {@inheritdoc} */
+	public function getFooterScripts(App\Request $request)
+	{
+		$moduleName = $request->getModule();
```
Comment (rskrzypczak) [inline modules/Vtiger/views/Tiles.php]: parent::

File: public_html/layouts/basic/modules/Vtiger/resources/List.js [hunk 15]
```diff
@@ -1238,6 +1243,11 @@ $.Class(
 				this.breadCrumbsFilter(selectOption.text());
 				this.ListViewPostOperation();
 				this.updatePagination(1);
+				if ('Tiles' === app.getViewName()) {
```
Comment (rskrzypczak) [inline public_html/layouts/basic/modules/Vtiger/resources/List.js]: ?

File: public_html/layouts/basic/modules/Vtiger/resources/Tiles.js [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+	{},
+	{
+		contentContainer: false,
+		/**
+		 * Register tiles size change action
+		 * @param {jQuery} topMenuContainer
+		 * @param {Vtiger_List_Js} listInstance
+		 */
+		registerTileSizeChange: function (topMenuContainer, listInstance) {
+			let thisInstance = this;
```
Comment (rskrzypczak) [inline public_html/layouts/basic/modules/Vtiger/resources/Tiles.js]:
```suggestion

```

File: public_html/layouts/basic/modules/Vtiger/resources/Tiles.js [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+					search_params: '',
+					advancedConditions: '',
+					tile_size: $(e.currentTarget).find('option:selected').val()
+				};
+				$('#recordsCount').val('');
+				$('#totalPageCount').text('');
+				topMenuContainer.find('.pagination').data('totalCount', 0);
+				listInstance.getListViewRecords(urlParams).done(() => {
+					listInstance.updatePagination(1);
+					thisInstance.setHeightOfTiles(thisInstance.contentContainer);
```
Comment (rskrzypczak) [inline public_html/layouts/basic/modules/Vtiger/resources/Tiles.js]:
```suggestion
					this.setHeightOfTiles(this.contentContainer);
```

File: public_html/layouts/basic/modules/Vtiger/resources/Tiles.js [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+				});
+				e.stopPropagation();
+			});
+		},
+		/**
+		 * Adjust height tile to the highest
+		 */
+		setHeightOfTiles: function () {
```
Comment (rskrzypczak) [inline public_html/layouts/basic/modules/Vtiger/resources/Tiles.js:41]: ?

File: public_html/layouts/basic/styles/components/_Global.scss [hunk 19]
```diff
@@ -677,3 +677,43 @@ input[type="checkbox"][disabled] {
...
+.js-tiles-container .c-tile-container:hover {
+	z-index: 999999;
+}
+
+.c-tile-body .c-multi-image__preview-img:hover,
+.c-tile-image:hover {
+	-moz-transform: scale(1.5);
+	-webkit-transform: scale(1.5);
+	transform: scale(1.5);
+	z-index: 999;
```
Comment (rskrzypczak) [inline public_html/layouts/basic/styles/components/_Global.scss]:
```suggestion

```

**Review by rskrzypczak — CHANGES_REQUESTED** [review]

File: config/ConfigTemplates.php [hunk 20]
```diff
@@ -600,6 +600,10 @@
 			'validation' => '\App\Validator::bool',
 			'sanitization' => '\App\Purifier::bool',
 		],
+		'tileDefaultSize' => [
+			'default' => 'very_small',
+			'description' => 'Should the breadcrumb menu have href enabled? true = enabled, false = off',
```
Comment (rskrzypczak) [inline config/ConfigTemplates.php]: Desc

File: layouts/basic/modules/Vtiger/TilesActions.tpl [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+						{include file=\App\Layout::getTemplatePath('ButtonLink.tpl', $MODULE) BUTTON_VIEW='listViewBasic'}
+					{/foreach}
+				{else}
+					<div class="dropright u-remove-dropdown-icon">
+						<button class="btn btn-xs  toolsAction dropdown-toggle" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+							<span class="fas fa-wrench" title="{\App\Language::translate('LBL_ACTIONS')}"></span>
+						</button>
+						<div class="dropdown-menu p-1" aria-label="{\App\Language::translate('LBL_ACTIONS')}">
+							{foreach from=$LINKS item=LINK}
+								{include file=\App\Layout::getTemplatePath('ButtonLink.tpl', $MODULE) BUTTON_VIEW='listViewBasic' BTN_CLASS='btn-xs'}
```
Comment (rskrzypczak) [inline layouts/basic/modules/Vtiger/TilesActions.tpl]:
```suggestion
								{include file=\App\Layout::getTemplatePath('ButtonLink.tpl', $MODULE_NAME) BUTTON_VIEW='listViewBasic' BTN_CLASS='btn-xs'}
```

File: layouts/basic/modules/Vtiger/TilesContents.tpl [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+								{include file=\App\Layout::getTemplatePath($FIELD_UI_TYPE_MODEL->getListSearchTemplateName(), $MODULE_NAME) FIELD_MODEL=$LISTVIEW_HEADER SEARCH_INFO=$SEARCH_INFO USER_MODEL=$USER_MODEL}
+							</td>
+						{/foreach}
+						<td class="reducePadding"></td>
+					</tr>
+				{/if}
+			</thead>
+			<tbody>
+			</tbody>
+		</table>
```
Comment (rskrzypczak) [inline layouts/basic/modules/Vtiger/TilesContents.tpl]: table move to the new shared file

File: layouts/basic/modules/Vtiger/TilesContents.tpl [hunk 23]
```diff
@@ -0,0 +1,179 @@
...
+								{include file=\App\Layout::getTemplatePath('TilesActions.tpl', $MODULE_NAME)}
+							</div>
+							<div class="card-body js-card-body justify-content-center h-100">
+								{foreach item=LISTVIEW_HEADER from=$LISTVIEW_HEADERS name=listHeaderForeach}
+									{assign var="FIELD_DATA_TYPE" value=$LISTVIEW_HEADER->getFieldDataType() }
+									{if $smarty.foreach.listHeaderForeach.first}
+										<h5 class="card-title text-center c-tile-value {if in_array($FIELD_DATA_TYPE,['multiImage', 'image'])} c-tile-image {/if}"><span class=" listViewEntryValue noWrap text-muted" data-field-type="{$FIELD_DATA_TYPE}">
+												{if empty($LISTVIEW_HEADER->get('source_field_name')) && ($LISTVIEW_HEADER->isNameField() eq true or $LISTVIEW_HEADER->getUIType() eq '4') && $MODULE_MODEL->isListViewNameFieldNavigationEnabled() eq true && $LISTVIEW_ENTRY->isViewable()}
+													<a {if $LISTVIEW_HEADER->isNameField() eq true}class="modCT_{$MODULE} js-list-field js-popover-tooltip--record" data-js="width" {/if} href="{$LISTVIEW_ENTRY->getDetailViewUrl()}">
+														<small> {$LISTVIEW_ENTRY->getListViewDisplayValue($LISTVIEW_HEADER)} </small>
```
Comment (rskrzypczak) [inline layouts/basic/modules/Vtiger/TilesContents.tpl]: ?

File: modules/Vtiger/models/Record.php [hunk 24]
```diff
@@ -480,6 +480,28 @@ public function getListViewDisplayValue($field, $rawText = false)
...
 
+	/**
+	 * Function to get the display value in Tiles.
+	 *
+	 * @param string|Vtiger_Field_Model $field
+	 * @param bool                      $rawText
+	 *
+	 * @throws \App\Exceptions\AppException
+	 *
+	 * @return mixed
```
Comment (rskrzypczak) [inline modules/Vtiger/models/Record.php]:
```suggestion
	 * @return string
```

File: modules/Vtiger/models/Record.php [hunk 25]
```diff
@@ -480,6 +480,28 @@ public function getListViewDisplayValue($field, $rawText = false)
...
+	 */
+	public function getTilesDisplayValue($field, $rawText = false)
+	{
+		if ($field instanceof Vtiger_Field_Model) {
+			if (!empty($field->get('source_field_name')) && isset($this->ext[$field->get('source_field_name')][$field->getModuleName()])) {
+				return $this->ext[$field->get('source_field_name')][$field->getModuleName()]->getListViewDisplayValue($field, $rawText);
```
Comment (rskrzypczak) [inline modules/Vtiger/models/Record.php]: ?

**Review by rskrzypczak — CHANGES_REQUESTED** [review]

File: modules/Vtiger/uitypes/Base.php [hunk 26]
```diff
@@ -232,6 +232,21 @@ public function getListViewDisplayValue($value, $record = false, $recordModel =
...
 
+	/**
+	 * Function to get the tile value in display view.
+	 *
+	 * @param mixed                    $value
+	 * @param bool|int                 $record
+	 * @param bool|Vtiger_Record_Model $recordModel
+	 * @param bool                     $rawText
+	 *
+	 * @return mixed
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/Base.php]:
```suggestion
	 * @return string
```

File: modules/Vtiger/uitypes/MultiImage.php [hunk 27]
```diff
@@ -440,4 +440,58 @@ public function uploadTempFile(array $files, int $recordId, ?string $hash = null
 		}
 		return $attach;
 	}
+
+	/** {@inheritdoc} */
+	public function getTilesDisplayValue($value, $record = false, $recordModel = false, $rawText = false)
+	{
+		if (!($value = \App\Json::decode($value)) && !$value) {
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/MultiImage.php]:
```suggestion
		if (!$value || !($value = \App\Json::decode($value))) {
```

File: modules/Vtiger/uitypes/MultiImage.php [hunk 28]
```diff
@@ -440,4 +440,58 @@ public function uploadTempFile(array $files, int $recordId, ?string $hash = null
...
+			}
+			return \App\Purifier::encodeHtml(trim($result, "\n\t ,"));
+		}
+
+		if (!\is_array($value)) {
+			$result .= '</div></div>';
+			return $result;
+		}
+		if ($record) {
+			$carouselId = App\Layout::getUniqueId("IC{$record}");
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/MultiImage.php]: ?

File: modules/Vtiger/uitypes/MultiImage.php [hunk 29]
```diff
@@ -440,4 +440,58 @@ public function uploadTempFile(array $files, int $recordId, ?string $hash = null
...
+			$result = '<div id="' . $carouselId . '" class="carousel slide" data-ride="carousel">
+		<div class="carousel-inner">';
+			foreach ($value as $itemNumber => $item) {
+				if ($record) {
+					$active = 0 === $itemNumber ? 'active' : '';
+					$result .= '    <div class="carousel-item ' . $active . '">
+				<img class="d-block  carousel-image" src="' . $this->getImageUrl($item['key'], $record) . '" alt="First slide">
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/MultiImage.php]: remove unnecessary spaces

File: public_html/layouts/basic/modules/Vtiger/resources/List.js [hunk 30]
```diff
@@ -1193,6 +1193,7 @@ $.Class(
 		 * Function to register the event for changing the custom Filter
 		 */
 		registerChangeCustomFilterEvent: function (event) {
+			const thisInstance = this;
```
Comment (rskrzypczak) [inline public_html/layouts/basic/modules/Vtiger/resources/List.js]: ?

File: public_html/layouts/basic/modules/Vtiger/resources/Tiles.js [hunk 31]
```diff
@@ -0,0 +1,80 @@
...
+				let urlParams = {
+					viewname: listInstance.getCurrentCvId(),
+					search_key: listInstance.getAlphabetSearchField(),
+					search_value: '',
+					search_params: '',
+					advancedConditions: '',
+					tile_size: selectedTileSize
+				};
+				$('#recordsCount').val('');
+				$('#totalPageCount').text('');
```
Comment (rskrzypczak) [inline public_html/layouts/basic/modules/Vtiger/resources/Tiles.js]: container

**Review by rskrzypczak — CHANGES_REQUESTED** [review]

File: layouts/basic/modules/Vtiger/TilesSize.tpl [hunk 32]
```diff
@@ -0,0 +1,18 @@
...
+	<!-- tpl-Base-TileSize -->
+	<div class="d-inline-block mr-sm-1 mb-1 mb-sm-0 c-btn-block-sm-down">
+		<button class="btn btn-light dropdown-toggle" data-toggle="dropdown" aria-expanded="true">
+			<span class="fas fa-th-large"></span>
+			<span class="textHolder ml-2 js-tile-dropdown-title"> {App\Language::translate('LBL_TILE_'|cat:$TILE_SIZE|upper, $MODULE_NAME)}</span>
+			<span class="caret"></span>
+		</button>
+		<div class="dropdown-menu js-selected-tile-size" data-selected-tile-size="{$TILE_SIZE}">
+			<a class="dropdown-item js-tile-size" data-tile-size="very_small" href="#">{\App\Language::translate('LBL_TILE_VERY_SMALL', $MODULE_NAME)}</a>
+			<a class=" dropdown-item js-tile-size" data-tile-size="small" href="#">{\App\Language::translate('LBL_TILE_SMALL', $MODULE_NAME)}</a>
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/TilesSize.tpl]:
```suggestion
			<a class="dropdown-item js-tile-size" data-tile-size="small" href="#">{\App\Language::translate('LBL_TILE_SMALL', $MODULE_NAME)}</a>
```

File: public_html/layouts/basic/styles/Main.scss [hunk 33]
```diff
@@ -39,6 +39,7 @@
 @import "pages/_List";
 @import "pages/_ListPreview";
 @import "pages/_Kanban";
+@import "pages/Tiles";
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/styles/Main.scss]:
```suggestion
@import "pages/_Tiles";
```

**Pull request merged**